visualization/view_scene.py

```diff
@@ -0,0 +1,111 @@
+"""
+Script to render a GTSFM scene using either Open3d or Mayavi mlab.
+
+Author: John Lambert
+"""
+
+import argparse
+import os
+from pathlib import Path
+
+import numpy as np
+from gtsam import Rot3, Pose3
+
+import gtsfm.utils.io as io_utils
+
+from visualization.open3d_vis_utils import draw_scene_open3d
+from visualization.mayavi_vis_utils import draw_scene_mayavi
+
+REPO_ROOT = Path(__file__).parent.parent.resolve()
+
+
+def compute_point_cloud_center_robust(point_cloud: np.ndarray) -> np.ndarray:
+    """Robustly estimate the point cloud center.
+
+    Args:
+        point_cloud: array of shape (N,3) representing 3d points.
+
+    Returns:
+        mean_pt: coordinates of central point, ignoring outliers.
+    """
+    ranges = np.linalg.norm(point_cloud, axis=1)
+    outlier_thresh = np.percentile(ranges, 75)
+    mean_pt = point_cloud[ranges < outlier_thresh].mean(axis=0)
+    return mean_pt
+
+
+def view_scene(args: argparse.Namespace) -> None:
+    """Read GTSFM output from .txt files and render the scene to the GUI.
+
+    Args:
+        args: rendering options.
+    """
+    points_fpath = f"{args.output_dir}/points3D.txt"
+    images_fpath = f"{args.output_dir}/images.txt"
+    cameras_fpath = f"{args.output_dir}/cameras.txt"
+
+    wTi_list, img_fnames = io_utils.read_images_txt(images_fpath)
+    calibrations = io_utils.read_cameras_txt(cameras_fpath)
+
+    if len(calibrations) == 1:
+        calibrations = calibrations * len(img_fnames)
+
+    point_cloud, rgb = io_utils.read_points_txt(points_fpath)
+
+    mean_pt = compute_point_cloud_center_robust(point_cloud)
+
+    # Zero-center the point cloud (about estimated center)
+    zcwTw = Pose3(Rot3(np.eye(3)), -mean_pt)
+    # expression below is equivalent to applying zcwTw.transformFrom() to each world point
+    point_cloud -= mean_pt
+
+    is_nearby = np.linalg.norm(point_cloud, axis=1) < args.max_range
+    point_cloud = point_cloud[is_nearby]
+    rgb = rgb[is_nearby]
+
+    if args.rendering_library == "open3d":
+        draw_scene_open3d(point_cloud, rgb, wTi_list, calibrations, zcwTw, args)
+    elif args.rendering_library == "mayavi":
+        draw_scene_mayavi(point_cloud, rgb, wTi_list, calibrations, zcwTw, args)
+    else:
+        raise RuntimeError("Unsupported rendering library")
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description="Visualize GTSFM result with Mayavi or Open3d.")
+    parser.add_argument(
+        "--rendering_library",
+        type=str,
+        default="mayavi",
+        choices=["mayavi", "open3d"],
+        help="3d rendering library to use.",
+    )
+    parser.add_argument(
+        "--output_dir", type=str, default=os.path.join(REPO_ROOT, "results", "ba_output"), help="Path to GTSFM output"
```

Comment:
I think we should take the full path of the GTSFMData object as input. I don't find `output` term as a clear one.

Reply:
So we use 3 file paths inside this directory, to create the GTSFMData. I think providing 1 directory is easier than providing the 3 file paths:
```
    points_fpath = f"{args.output_dir}/points3D.txt"
    images_fpath = f"{args.output_dir}/images.txt"
    cameras_fpath = f"{args.output_dir}/cameras.txt"
```
Maybe I should clarify the argparse docstring for this arg more @ayushbaid ?

Reply:
Makes sense. Maybe use '--gtsfm-data-dir' or '--scene-data-dir'?

Reply:
so when I think of "data dir", I think of the input image directory. And the same for scene data dir. But I cleaned up the `help` description. Let me know if this is clearer now.